syntax = "proto3";

package bloock;

option go_package = "github.com/bloock/bloock-sdk-go/v2/internal/bridge/proto";
option java_package = "com.bloock.sdk.bridge.proto";

import "keys_entities.proto";

enum ProofType {
    IntegrityProofType = 0;
    SparseMtProofType = 1;
}

message IssuerKey {
    optional LocalKey local_key = 1;
    optional ManagedKey managed_key = 2;
}

message CredentialV2 {
    repeated string context = 1;
    string id = 2;
    repeated string type = 3;
    string issuance_date = 4;
    string expiration = 5;
    string credential_subject = 6;
    CredentialStatusV2 credential_status = 7;
    string issuer = 8;
    CredentialSchemaV2 credential_schema = 9;
    CredentialProofV2 proof = 10;
}

message CredentialProofV2 {
    string signature_proof = 1;
    optional string integrity_proof = 2;
    optional string sparse_mt_proof = 3;
}

message CredentialStatusV2 {
    string id = 1;
    int64 revocation_nonce = 2;
    string type = 3;
}

message CredentialSchemaV2 {
    string id = 1;
    string type = 2;
}

message StringAttributeV2 {
    string id = 1;
    string value = 2;
}

message IntegerAttributeV2 {
    string id = 1;
    int64 value = 2;
}

message DecimalAttributeV2 {
    string id = 1;
    double value = 2;
}

message BooleanAttributeV2 {
    string id = 1;
    bool value = 2;
}

message DateAttributeV2 {
    string id = 1;
    string value = 2;
}

message DateTimeAttributeV2 {
    string id = 1;
    string value = 2;
}

message StringAttributeDefinitionV2 {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message IntegerAttributeDefinitionV2 {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message DecimalAttributeDefinitionV2 {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message BooleanAttributeDefinitionV2 {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message DateAttributeDefinitionV2 {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message DateTimeAttributeDefinitionV2 {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
}

message StringEnumAttributeDefinitionV2 {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
    repeated string enum = 5;
}

message IntegerEnumAttributeDefinitionV2 {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
    repeated int64 enum = 5;
}

message DecimalEnumAttributeDefinitionV2 {
    string display_name = 1;
    string id = 2;
    string description = 3;
    bool required = 4;
    repeated double enum = 5;
}

message CredentialReceiptV2 {
    CredentialV2 credential = 1;
    string credential_id = 2;
    string credential_type = 3;
    optional int64 anchor_id = 4;
}

message IssuerStateReceipt {
    string tx_hash = 1;
}

message SchemaV2 {
    string cid = 1;
    string cid_json_ld = 2;
    string json = 3;
}

message CredentialRevocationV2 {
    bool success = 1;
}

message IssuerParams {
    Method method = 1;
    Blockchain blockchain = 2;
    NetworkId network_id = 3;
}

enum Method {
    IDEN3 = 0;
    POLYGON_ID = 1;
}

enum Blockchain {
    ETHEREUM = 0;
    POLYGON = 1;
    UNKNOWN_CHAIN = 2;
    NO_CHAIN = 3;
}

enum NetworkId {
    MAIN = 0;
    MUMBAI = 1;
    GOERLI = 2;
    UNKNOWN_NETWORK = 3;
    NO_NETWORK = 4;
}